### DISCUSSION TITLE
Impulse invariant discretization function for continuous transfer functions from physics.control.lti

### DISCUSSION THREAD

**Discussion by zolabar** [sympy/sympy#24643] [Ideas]
The bilinear discretization transform for a continuous transfer function $H(s)=\frac{b_0+b_1 s+b_2 s^2+...}{a_0+b_1 s+a_2 s^2+...}$  discussed in

[issue-24544 on bilinear discretization](https://github.com/sympy/sympy/issues/24544)

and

[pr-24558](https://github.com/sympy/sympy/pull/24558)

is frequency-response-invariant and stability preserving.

https://en.wikipedia.org/wiki/Bilinear_transform

There are numerical counterparts

https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.bilinear.html

https://de.mathworks.com/help/signal/ref/bilinear.html

In digital circuit design a impulse response invariance may be preferred to a frequency-response-invariance.

https://en.wikipedia.org/wiki/Impulse_invariance

A numerical counterpart exists in 

https://de.mathworks.com/help/signal/ref/impinvar.html

Here is a proposition of a symbolic impulse invariant transformation, in analogy to the bilinear discretization, based on partial fraction decomposition. 

[pr-24558](https://github.com/sympy/sympy/pull/24558)

```
import sympy as sym
from sympy.physics.control.lti import TransferFunction

def impinvar(tf, sample_per):
    num, den = tf.num, tf.den
    s = tf.var
    roots = list(sym.roots(den, s))
    main = den.as_poly(s).all_coeffs()[0]
    pfd = 0
    discrete_var = sym.Symbol('z')
    orderNum = len(tf.num.as_poly(s).all_coeffs())-1
    orderDen = len(tf.den.as_poly(s).all_coeffs())-1

    if not orderDen-orderNum>0:
        print()
        print('impinvar works only for strictly proper transfer functions, i.e. order of den strictly greater than order of num')
        print()
        return
    
    for i in range(0, len(roots)):
        dummy_den = 1
        for j in range(0, len(roots)):
            if j is not i:
                dummy_den = dummy_den * (s-roots[j])
                                         
        dummy_coef = (tf.num/dummy_den).subs(s, roots[i])/main                                         
        pfd = pfd + dummy_coef/(discrete_var-sym.exp(roots[i]*sample_per))
    
    pfd = sample_per*discrete_var*pfd
    
    
    if len(roots) > 1:
        #print(len(roots))
        pfd = sym.ratsimp(pfd).simplify()
    
    numZ, denZ = sym.fraction(pfd)
    
    numZ = numZ.as_poly(discrete_var)
    denZ = denZ.as_poly(discrete_var)
    
    
    #print(sym.latex(numZ.as_expr()/denZ.as_expr()))
    
    return numZ.all_coeffs(), denZ.all_coeffs()    
  

```

example

```
impinvar(TransferFunction(1, s, s), T)
([T, 0], [1, -1])
```

@hanspi42, what do you think of the `impinvar` function? and of the following partial fraction decomposition?

One could also add the partial fraction decomposition to physics.control.lti

```
def partial_fraction(tf):
    num, den = tf.num, tf.den
    s = tf.var
    main = den.as_poly(s).all_coeffs()[0]
    roots = list(sym.roots(tf.den, s))
    print(roots)
    pfd = 0

    for i in range(0, len(roots)):
        dummy_den = 1
        for j in range(0, len(roots)):
            if j is not i:
                dummy_den = dummy_den * (s-roots[j])
                                         
        dummy_coef = (tf.num/dummy_den).subs(s, roots[i])/main                                         
        pfd = pfd + dummy_coef/(s-roots[i])
 
    
    return pfd

lewis = TransferFunction(16*s+16, s**3+6*s**2 +8*s, s)
partial_fraction(lewis) 
```
```
impinvar(lewis, 0.1)
([0.125300273609809, -0.113348106213728, 0.0],
 [1.00000000000000, -2.48905079911362, 2.03786243520765, -0.548811636094026])
```

**Comment by hanspi42** [comment]
> @hanspi42, what do you think of the `impinvar` function? and of the following partial fraction decomposition?
> 
> One could also add the partial fraction decomposition to physics.control.lti

We already have partial fraction decomposition in SymPy. Try

```
(lewis.num/lewis.den).apart(s)
```

If you need the decomposition to single poles,

```
(lewis.num/lewis.den).apart(s, full=True).doit()
```

What would you do with multiple poles, by the way?

**Reply by zolabar** [reply]
@hanspi42 , thank you for the comnents, i wanted to use them in the impinvar function, that calculates the discrte coeffs of a cont tf discretized with the impulse invariant method

**Reply by hanspi42** [reply]
I know. I mean: what do you do if you have multiple poles like in `s/(s+1)**3`?

**Reply by zolabar** [reply]
Ah, I see… I didn‘t think of that!

**Comment by zolabar** [comment]
Since to my current knowledge, the substitution discretizations backward_diff, bilinear and forward_diff, seem practical for usage I don‘t know how relevant the impulse invariant, step invariant, etc… discretization methods are…
